proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+    (Ditto ``ghc-prim``.) Need to make sure that if someone wants to follow the Haddock source-code link to (say) Functor, they should still find it regardless of where it is actually defined.
+
+  - We could consider issuing a warning if you say ``-package ghc-internals`` (or ``ghc-bignum`` or ``ghc-prim``), one that was hard to silence.
+    Since we can have module-level ``WARNING`` pragmas with custom categories, one way to realise this would be to pick a category and add such pragmas to every module in the relevant packages, though we might want to do something more systematic.
+    The text of the warnings could encourage users to
+
+    - switch to a function exposed by base, and/or
+    - petition the CLC to expose this super-useful function from base.
+
+  - ``cabal check`` (a per-package check) could warn on packages that use ``ghc-internals``.
+
+  - ...what else?
+
+- In contrast, clients are *not* discouraged from depending on ``ghc-experimental``; although again its name should convey the idea that it might change at short notice.
+
+  ``ghc-experimental`` allows the GHC Steering Committee to make initially-experimental language changes, which often involve new types and functions, without committing to permanently supporting the precise API, since it often takes a little while for these designs to settle down.
+
+  The existence of ``ghc-experimental`` should substantially ameliorate the difficulty that many GHC Proposals have a library-function component, but it is unlikely to be a *stable* API (having just been invented) and is therefore in conflict with the CLC's goals.
+
+  As they become stable, the CLC may want to consider adopting the new types and functions from ``ghc-experimental`` into ``base``.
+  (But CLC would not expect to curate the API of ``ghc-experimental``.)
+
+- Perhaps ``ghc-experimental`` should be in the purview of the GHC Proposals process.
+  GHC devs should not just make up random APIs and pop them into ``ghc-experimental``; a scrutiny process would be valuable.
+
+- Under this proposal, there is initially no change (whatsoever) to the API exposed by ``base``, or its performance characteristics.
+  The impact on clients should therefore be zero.
+
+  Over time, the GHC developers may make CLC proposals to remove types and functions that are currently in the ``base`` API, but are in truth part of GHC's implementation, and were originally exposed by historical accident.
+  But these are *future*\ proposals.
+
+  To make the transition suggested in these future proposals easier to manage, we have in progress a "deprecated exports" mechanism that will ease such transitions.
+  For a transitional period, ``base`` can continue to export the function, but with a deprecation warning saying something like:
+
+    This is going to disappear from base.
+    You probably don't want to use it at all.
+    But if you absolutely must, get it from ``ghc-internals``.
```

Comment:
We can't say in the same document that we should use "all means necessary" to avoid `ghc-internals` dependencies, and also recommend people to import a function from there.

```suggestion
    If you absolutely must, define it yourself locally, or if this is impossible, let the CLC and GHC team know!
```

Reply:
I would expect `ghc-internals` to mainly contain things one cannot define locally? I would expect also some zany integrate-with-ghc-so-tightly-someone-turns-blue projects to always need to use `ghc-internals` too. I think it's less important to deny that category exists, than emphasize that the vast majority of projects are *not* in that category.

Put another way, I'd expect those projects to know who they are and not reach out to GHC devs or the CLC, and I'd expect almost everyone who does reach out to be told "actually, you don't need to do things that way". Is that what you were thinking too? Or were you imagining we might sometimes move something to `ghc-internals` (without a reexport elsewhere) that shouldn't be there, by mistake?

Reply:
If such projects are relying on functionality from `base` that GHC developers and the CLC both agree should be moved to `ghc-internals`, they ought to reach out and say "hey, we use this for such-and-such", and make a case for it to be put in `ghc-prim` instead.

The point here is to make, FAIAP, a "hidden" library containing code in the strictest intersection of "compiler bootstrap" and "necessary to allow general-purpose programming": stuff like the guts of IO, datatype representations, and the like. There should be nothing  in `ghc-internals` that is both:
- useful for general programming, and
- not exported from `base`, `ghc-prim`, or `ghc-experimental`.

Reply:
I agree with the main "There should be nothing in ghc-internals that is both: useful for general programming, and not exported from..." point, but do not that I don't think anyone intends `ghc-prim` to be any more user-facing than `ghc-internals`. Perhaps the proposal should make this more explicit?

Reply:
> not exported from base, ghc-prim, or ghc-experimental.

This should be "not exported fomr base or ghc-experimental".  `ghc-prim` is every bit "internal" as `ghc-internals`.   Anything in `ghc-prim` that is useful for general programming should be be exposed by `base` or `ghc-experimenal`.  Indeed maybe we should make that clearer.